{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/env python\n",
    "# coding: utf-8\n",
    "\n",
    "# In[7]:\n",
    "\n",
    "\n",
    "import os\n",
    "import re\n",
    "import sys\n",
    "import subprocess\n",
    "from Bio import SeqIO\n",
    "\n",
    "#We make genes an empty list\n",
    "genes = []\n",
    "\n",
    "\n",
    "working_directory = sys.argv[1]\n",
    "os.mkdir(working_directory + \"/Alignment_Files\")\n",
    "\n",
    "#HAVENT EVEN USED THE PARAGRAPH BELOW\n",
    "#For the used geneList file, we grab the gene names!\n",
    "for lines in open(working_directory + \"/Tree_Genes/geneList.fasta\"):\n",
    "    #Strip the new lines\n",
    "    line = lines.rstrip(\"\\n\")\n",
    "    #And look for the > to indicate gene name\n",
    "    if re.search(\">\",line):\n",
    "        #Then replace > with nothing to obtain just the gene name\n",
    "        line = re.sub(\">\",\"\",line)\n",
    "        #Finally, we have the gene name\n",
    "        genes.append(line)\n",
    "\n",
    "#Now we need to extract the sequence matches from the BlastResults file and concatenate them based on gene type\n",
    "\n",
    "#For each file in the directory BlastResults (which is each strains sequence matches to the tree genes)\n",
    "strains = []\n",
    "for file in os.listdir(working_directory + \"/BlastResults/\"):\n",
    "    #print(file)\n",
    "    strain = re.sub(\".fasta\",\"\",file)\n",
    "    strains.append(strain)\n",
    "    #We create a new dict called sequences\n",
    "    sequences = {}\n",
    "    #We then open the file to view all the lines\n",
    "    for lines in open(working_directory + \"/BlastResults/\" + file):\n",
    "        #We strip the newline\n",
    "        line = lines.rstrip(\"\\n\")\n",
    "        #Indicate those gene names beginning with > as the key value\n",
    "        if re.search(\">\", line):\n",
    "            key = line\n",
    "            #print(key)\n",
    "        else:\n",
    "            #And those that do not (i.e., sequence of the gene) as the value in the dict\n",
    "            value = line\n",
    "            #print(value)\n",
    "            #And combine both to a dict i.e. Gene: Sequence\n",
    "            sequences[key] = value\n",
    "    #print(sequences)\n",
    "    \n",
    "    #For each gene found in the sequence for a specific file\n",
    "    for genes_present in sequences.keys():\n",
    "        file_name = re.sub(\">\",\"\",genes_present)\n",
    "        #We open a file called the gene name (which is appendable, so we don't overwrite)\n",
    "        with open(working_directory+ \"/Alignment_Files/\" + file_name + \".tsv\", 'a') as nfile:\n",
    "            #And save the sequence for each gene, in FASTA format, using the strain name as the header for each sequence.\n",
    "            nfile.write(\">\" + file + \"\\n\" + sequences[genes_present] + \"\\n\")\n",
    "\n",
    "#Come back to below, this is a method to run bash commands through python (but it doesn't work on my windows computer)\n",
    "bashCommand = \"for x in ../Alignment_Files/*.tsv; do mafft $x > ../Alignment_Files/${x##*/}_aligned.csv; done\"\n",
    "process = subprocess.Popen(bashCommand, stdout=subprocess.PIPE, shell=True, executable='/bin/bash')\n",
    "output, error = process.communicate()\n",
    "\n",
    "#This is how to generate dashes for each gene type based on aligned sequences\n",
    "#This is for the strains that had no matching sequence to a particular gene.\n",
    "#Set up a dict to store the gene name and the corresponding dash sequence\n",
    "each_file_dashed = {}            \n",
    "#Open each file, as long as it's a csv file (will be changed for final script)\n",
    "for file in os.listdir(working_directory + \"/Alignment_Files/\"):\n",
    "    if file.endswith(\".csv\"):\n",
    "        for lines in open(working_directory + \"/Alignment_Files/\" + file):\n",
    "            line = lines.rstrip(\"\\n\")\n",
    "            #If a header, then restart dashes amount\n",
    "            if line.startswith(\">\"):\n",
    "                #print(line)\n",
    "                dashes = \"\"\n",
    "                key = file\n",
    "            else:\n",
    "                dashes = dashes + (len(line) * \"-\")\n",
    "        #print(dashes)\n",
    "        file = re.sub(\".tsv_aligned.csv\",\"\",file)\n",
    "        each_file_dashed[file] = dashes\n",
    "\n",
    "#d = SeqIO.to_dict(SeqIO.parse(\"../py/>chuA.tsv_aligned.csv\", 'fasta'))\n",
    "\n",
    "#print(d)\n",
    "\n",
    "\n",
    "for strain in strains:\n",
    "    sequence = \"\"\n",
    "    for file in os.listdir(working_directory + \"/Alignment_Files/\"):\n",
    "        if file.endswith(\".csv\"):\n",
    "            d = SeqIO.to_dict(SeqIO.parse(working_directory + '/Alignment_Files/' + file, 'fasta'))\n",
    "            #strain = strain + \".fasta\"\n",
    "            try:\n",
    "                header = re.sub(\".tsv_aligned.csv\",\"\",file)\n",
    "                sequence = sequence + str(d[strain+'.fasta'].seq)\n",
    "                #with open(strain+\"_final.tsv\", 'a') as nfile:\n",
    "                    #And save the sequence for each gene, in FASTA format, using the strain name as the header for each sequence.\n",
    "                    #nfile.write(header + \"\\n\" + str(d[strain+'.fasta'].seq) + \"\\n\")\n",
    "                    \n",
    "                    #print(strain)\n",
    "                    #print(\"success\")\n",
    "                    #print(header)\n",
    "                    #print(str(d[strain+'.fasta'].seq))\n",
    "            except:\n",
    "                #print(strain)\n",
    "                header = re.sub(\".tsv_aligned.csv\",\"\",file)\n",
    "                sequence = sequence + each_file_dashed[header]\n",
    "                #print(\"failure\")\n",
    "                #print(header)\n",
    "                #with open(strain+\"_final.tsv\", 'a') as nfile:\n",
    "                    #And save the sequence for each gene, in FASTA format, using the strain name as the header for each sequence.\n",
    "                    #nfile.write(header + \"\\n\" + each_file_dashed[header] + \"\\n\")\n",
    "                #print(strain)\n",
    "                #print(each_file_dashed[\">aceK_arpA\"])\n",
    "    #print(strain + \"\\n\" + sequence)\n",
    "    with open(working_directory + \"/final_alignment.tsv\", 'a') as nfile:\n",
    "        nfile.write(\">\" + strain + \"\\n\" + sequence + \"\\n\")\n",
    "        \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
