{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "10b2bb23",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running through AF334679_1SIVsmm.fasta for reformatting\n",
      "Running through BIV.fasta for reformatting\n",
      "Running through CAEV.fasta for reformatting\n",
      "Running through EIAV.fasta for reformatting\n",
      "Running through FIV.fasta for reformatting\n",
      "Running through HIV-1-M.fasta for reformatting\n",
      "Running through HIV-1-O.fasta for reformatting\n",
      "Running through HIV-1-P.fasta for reformatting\n",
      "Running through HIV-2.fasta for reformatting\n",
      "Running through JDV.fasta for reformatting\n",
      "Running through SIVagm-s.fasta for reformatting\n",
      "Running through SIVagm-t.fasta for reformatting\n",
      "Running through SIVcpz.fasta for reformatting\n",
      "Running through SIVgor.fasta for reformatting\n",
      "Running through SIVmnd-2.fasta for reformatting\n",
      "Running through Visna.fasta for reformatting\n"
     ]
    }
   ],
   "source": [
    "#import modules needed\n",
    "import os\n",
    "import re\n",
    "import sys\n",
    "\n",
    "#So here we would hope to be doing a correction to the results of the blast so that we're just getting one \n",
    "#gene match back per queried gene.\n",
    "\n",
    "#Establish a list to store the file names in\n",
    "file_names = []\n",
    "\n",
    "#This allows user inputted genome and gene directories.\n",
    "#blast_directory = sys.argv[1]\n",
    "blast_directory = \"/media/sam/Expansion/Bioinformatics/Glasgow_Andy_Patricia_Tree/Version1/BlastResults\"\n",
    "\n",
    "#For each file in the blast directory, add to the file name list\n",
    "for file_name in os.listdir(blast_directory):\n",
    "    #print(file_name)\n",
    "    file_names.append(file_name)\n",
    "    \n",
    "#print(range(len(file_names)))\n",
    "\n",
    "#For each file (strain)\n",
    "for new in range(len(file_names)):\n",
    "    print(\"Running through \" + file_names[new] + \" for reformatting\")\n",
    "    \n",
    "    #Establish lists to append to per iteration\n",
    "    new_info = dict()\n",
    "    genes = []\n",
    "    sequences = []\n",
    "    percentages = []\n",
    "    \n",
    "    \n",
    "    #Open a new file to save output to - this is just a new file per strain\n",
    "    new_file = open(blast_directory + \"/new_\" + file_names[new], \"w\")\n",
    "    \n",
    "    \n",
    "    #print(file_names[new])\n",
    "    \n",
    "    #Open each file in directory and iterate through the lines\n",
    "    for lines in open(blast_directory + \"/\" + file_names[new]):\n",
    "        \n",
    "        #If the line has a \">\" then it must be the gene name\n",
    "        if re.search(\">\", lines):\n",
    "            #So we append it to the genes list\n",
    "            genes.append(lines.rstrip(\"\\n\"))\n",
    "            \n",
    "        #Else if the length of the line is greater than 10, then it must be the sequence, not the percentage\n",
    "        elif range(len(lines) > 10):\n",
    "            #print(lines)\n",
    "            #So we can append it to the sequences list\n",
    "            sequences.append(lines.rstrip(\"\\n\"))\n",
    "            \n",
    "        #Otherwise it's the percentage\n",
    "        else:\n",
    "            #So we can append it to the percentages list\n",
    "            percentages.append(lines.rstrip(\"\\n\"))\n",
    "\n",
    "    #print(genes)\n",
    "    #print(sequences)\n",
    "    #print(percentages)\n",
    "\n",
    "    #Now we're checking to see if the next gene in the sequence matches the gene we're currently at\n",
    "    #For each index in genes\n",
    "    for x in range(len(genes)):\n",
    "\n",
    "        #Iterate through x + 1 to the last gene\n",
    "        for y in range(x+1,len(genes)):\n",
    "            #print(y)\n",
    "            #If the gene name of x equals the gene name of y\n",
    "            if (genes[x] == genes[y]):\n",
    "                #If the number of genes of the same name is above 2\n",
    "                if (genes.count(genes[x]) > 2):\n",
    "                    #Then grab the index number of each (we make the assumption 3 matches will be enough)\n",
    "                    indecesthree = [i for i, e in enumerate(genes) if e == genes[x]]\n",
    "                    #Store each gene index\n",
    "                    first = indecesthree[0]\n",
    "                    second = indecesthree[1]\n",
    "                    third = indecesthree[2]\n",
    "                    \n",
    "                    #Now compare the percentage coverage of each match and save the one with the highest match as \n",
    "                    #the winning sequence match to use\n",
    "                    if percentages[first] > percentages[second]:\n",
    "                        if percentages[first] > percentages[third]:\n",
    "                            new_info[genes[x]] = sequences[first]\n",
    "                        else:\n",
    "                            new_info[genes[x]] = sequences[third]\n",
    "                    else:\n",
    "                        if percentages[second] > percentages[third]:\n",
    "                            new_info[genes[x]] = sequences[second]\n",
    "                        else:\n",
    "                            new_info[genes[x]] = sequences[third]\n",
    "\n",
    "               \n",
    "                #Now if we only have two of the same gene\n",
    "                elif (percentages[x] > percentages[y]):\n",
    "                    #Append the higher match sequence to the new_sequences\n",
    "                    new_info[genes[x]] = sequences[x]\n",
    "               \n",
    "                #Otherwise, y is less than x (or equal, which wouldn't matter which we chose)\n",
    "                else:\n",
    "                    #Append the sequence of y to the new_sequences\n",
    "                    new_info[genes[x]] = sequences[y]\n",
    "\n",
    "\n",
    "    #For each index in the genes\n",
    "    for x in range(len(genes)):\n",
    "        #If the gene name isn't in the new list, because it didn't have a match...\n",
    "        if genes[x] not in new_info:\n",
    "            #print(genes[x])\n",
    "            #We then add it to the new_genes list\n",
    "            new_info[genes[x]] = sequences[x]\n",
    "\n",
    "\n",
    "    \n",
    "    #And then we iteratively add the contents of both lists to the respective genome blast file match.\n",
    "    #for x in range(len(new_genes)):\n",
    "    #    new_file.write(new_genes[x] + \"\\n\" + new_sequences[x] + \"\\n\")\n",
    "    #print(new_info)\n",
    "    for x in new_info:\n",
    "        new_file.write(x + \"\\n\" + new_info[x] + \"\\n\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
